geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -146,21 +148,34 @@ object ScalaSimpleFeature {
     * @param readAttribute lazily read an attribute value, must already be converted into the appropriate types
     * @param readUserData lazily read the user data
     */
-  class LazyImmutableSimpleFeature(sft: SimpleFeatureType,
-                                   id: String,
-                                   readAttribute: Int => AnyRef,
-                                   readUserData: => java.util.Map[AnyRef, AnyRef])
-      extends AbstractImmutableSimpleFeature(sft, id) {
-
-    private val bits = AtomicBitSet(sft.getAttributeCount)
-    private val attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
-
-    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] =
-      Collections.unmodifiableMap(synchronized { readUserData })
+  class LazyImmutableSimpleFeature(
+      sft: SimpleFeatureType,
+      id: String,
+      private var readAttribute: Int => AnyRef,
+      private var readUserData: () => java.util.Map[AnyRef, AnyRef]
+    ) extends AbstractImmutableSimpleFeature(sft, id) {
+
+    private val bits = IntBitSet(sft.getAttributeCount)
+    private lazy val attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
+    private var count = 0
+
+    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] = {
+      val read = synchronized { readUserData() }
+      readUserData = null // dereference any backing resources
```

Comment:
Ok, I'll bite.  Can you explain this?

Reply:
once the user data has been deserialized, we set the read function to null, that way the garbage collector can reclaim the serialized byte array and any other intermediate constructs used for deserialization. otherwise we're hanging on to both the serialized bytes and the deserialized objects, so 2x the memory footprint.